reciprocalspaceship/io/crystfel.py

```diff
@@ -0,0 +1,190 @@
+                hkl = np.array([h, k, l])
+                # print(A)
+                s1 = A @ hkl + s0
+                s1x, s1y, s1z = s1
+                ewald_offset = np.linalg.norm(s1) - lambda_inv
+
+                record = {
+                    "H": h,
+                    "K": k,
+                    "L": l,
+                    "I": float(I),
+                    "sigmaI": float(sigmaI),
+                    "BATCH": crystal_idx,
+                    's1x': s1x,
+                    's1y': s1y,
+                    's1z': s1z,
+                    'ewald_offset': ewald_offset
+                }
+                if current_event is not None:
+                    name = (current_filename, current_event,
+                            current_serial_number, crystal_idx, crystal_peak_number)
+                else:
+                    name = (current_filename, current_serial_number, crystal_idx,
+                            crystal_peak_number)
+                answ_crystals[name] = record
+
+            # start analyzing where we are now
+            if corrupted_chunk:
+                if "Begin chunk" not in line:
+                    continue
+                else:
+                    is_crystal, is_chunk = False, False
+                    corrupted_chunk = False
+                    continue
+
+            if contains_filename(line):
+                current_filename = line.split()[-1]
+            elif contains_event(line):
+                current_event = line.split()[-1][2:]
+            elif contains_serial_number(line):
+                current_serial_number = line.split()[-1]
+
+            elif starts_chunk_peaks(line):
+                is_chunk = True
+                continue
+
+            elif starts_crystal_peaks(line):
+                crystal_idx += 1
+                is_crystal = True
+                continue
+
+    return answ_crystals
+
+
+def read_crystfel(streamfile, logfile=None) -> DataSet:
+    """
+    Initialize attributes and populate the DataSet object with data from a CrystFEL stream with indexed reflections. This is the output format used by CrystFEL software when processing still diffraction data.
+
+    Parameters
+    ----------
+    streamfile : str of file
+        name of a .stream file or a file object
+    """
+
+    if not streamfile.endswith('.stream'):
+        raise ValueError("Stream file should end with .stream")
+    # read data from stream file
+    d = _parse_stream(streamfile)
+    df = pd.DataFrame.from_records(list(d.values()))
+
+    # set mtztypes as in precognition.py
+    # hkl -- H
+    # I, sigmaI -- J, Q
+    # BATCH -- I
+    # s1{x,y,z} -- R
+    # ewald_offset -- R
+    names = [
+        'H', 'K', 'L', 'I', 'sigmaI', 'BATCH', 's1x', 's1y', 's1z',
+        'ewald_offset'
+    ]
+    mtztypes = ["H", "H", "H", "J", "Q", "I", "R", "R", "R", "R"]
+    dataset = DataSet()
+    for (k, v), mtztype in zip(df.items(), mtztypes):
+        dataset[k] = v
```

Comment:
To my knowledge, these two lines can be simplified to:
`dataset[k] = v.astype(mtztype)`

Reply:
I copied this line from `read_precognition.py` -- I believed you had some reason to do that :)

Reply:
fair enough -- it certainly works as written, so we can just leave it. I cleaned up that section of `read_precognition()` last night, and we can always apply the same changes here.